### PR TITLE
Run entire epoch as compiled tf.function in model.fit()

### PR DESCRIPTION
**Pull request by lgeiger**
As mentioned in #40881, this PR PR changes Keras `model.fit()` and `model.evaluate()` to automatically compile the iteration over one epoch into a `tf.function` to improve performance. This new fastpath will be triggered starting from the second epoch in the following cases:
- `run_eagerly` is `False`
- No batch level callback hooks need to be called (usually true when `verbose != 1`)
- `steps_per_execution == 1`
- `data_handler.inferred_steps` are defined

These changes should prevent users needing to learn about `experimental_steps_per_execution` and has the potential to greatly improve performance on GPUs or TPUs.

I briefly benchmarked these changes on a CPU only machine to make sure that this doesn't introduce any obvious regressions. @omalleyt12 I think you are in a much better place to judge this, so it would be great if you could run this against your internal benchmarks using TPUs or GPUs as well.

I think these changes should be covered by the current test cases, but please let me know if there is something I am missing here.

### CODE REVIEW

File: tensorflow/python/keras/engine/training.py [hunk 1]
```diff
@@ -1084,27 +1084,41 @@ def fit(self,
...
-                _r=1):
-              callbacks.on_train_batch_begin(step)
-              tmp_logs = self.train_function(iterator)
-              if data_handler.should_sync:
-                context.async_wait()
-              logs = tmp_logs  # No error, now safe to assign to logs.
-              end_step = step + data_handler.step_increment
-              callbacks.on_train_batch_end(end_step, logs)
+        if (run_compiled and data_handler.inferred_steps is not None and
+            epoch - initial_epoch > 0):
```
Comment (lgeiger) [inline tensorflow/python/keras/engine/training.py:1097]: We only use a `tf.function` for the second epoch as we need `logs` to be defined in order to prevent us needing to run the first train step outside the loop similar to:
https://github.com/tensorflow/tensorflow/blob/3a3d3591923501aa6fb03ef4da1d64400ad209d8/tensorflow/python/keras/engine/training.py#L813-L818

If I understand https://github.com/tensorflow/tensorflow/pull/40881#issuecomment-662731863 correctly this would otherwise lead to an increased graph size which can become problematic for larger networks.

In theory we could also only run the first training step outside of the `tf.function`, but I think this would make the code a bit more complicate and won't increase performance by much.

File: tensorflow/python/keras/engine/training.py [hunk 2]
```diff
@@ -1084,27 +1084,41 @@ def fit(self,
...
+          logs = epoch_train_fn(iterator, data_handler.inferred_steps, logs)
+        else:
+          with data_handler.catch_stop_iteration():
+            for step in data_handler.steps():
+              with trace.Trace(
+                  'train',
+                  epoch_num=epoch,
+                  step_num=step,
+                  batch_size=batch_size,
+                  _r=1):
```
Comment (lgeiger) [inline tensorflow/python/keras/engine/training.py:1114]: I didn't add tracing to the compiled code path as the `TensorBoard` profiling callback needs to have batch level hooks defined anyway so it won't be used in the compiled version anyway as far as I know.

**Comment by lgeiger** [comment]
@omalleyt12 Do you have time to take a look?

**Comment by lgeiger** [comment]
@omalleyt12 I rebase this PR onto master since AutoGraph now supports creating new symbols in a loop, the code could be simplified and support added for `model.evaluate()`.

Could you please take a look at this PR as it has to potential to greatly speedup training time in some cases without the users needing to worry about setting `steps_per_execution`.

File: tensorflow/python/keras/engine/training.py [hunk 3]
```diff
@@ -1347,19 +1369,24 @@ def evaluate(self,
       logs = {}
       self.test_function = self.make_test_function()
       self._test_counter.assign(0)
+      epoch_test_function = self._get_compiled_epoch_function(
```
Comment (lgeiger) [inline tensorflow/python/keras/engine/training.py:1372]: @omalleyt12 Do you think it would be useful to cache this function similar to what's done with `self._eval_data_handler`?

Comment (gbaned) [reply tensorflow/python/keras/engine/training.py:1372]: @omalleyt12 Can you please take a look on above comments from @lgeiger. Thanks!

**Comment by lgeiger** [comment]
@gbaned @omalleyt12 @fchollet Any news on this PR, it has been open for a long time and still awaits a PR review

**Comment by omalleyt12** [comment]
@lgeiger Thanks for the PR! Sorry for the delay in review!

Discussed with @fchollet , we think the best thing to do would be to "auto-tune" the `steps_per_execution` such that the timing of one execution takes > ~200ms-1s. This would achieve most of the benefits of running an entire epoch in one execution while still allowing the progress bar to update quickly in a Colab/Jupyter notebook and also ensuring that any error that occurs doesn't invalidate a whole epoch of training.

We should only do this in the case where (mostly as you said):

1. run_eagerly is False
2. No batch level callback hooks need to be called
3. User passes `steps_per_execution=None` (we should make sure a user can hard-code `steps_per_execution=1`, bc the way we do multiple steps still requires extra graph tracing which may cause OOM or performance issues)
4. `data_handler.inferred_steps` are defined

However, under this approach, it's ok to do for `verbose=1`, and we can exempt the `ProgressBarLogger` callback from criteria 2.

You've already done most of the leg-work for an approach like this in this PR, would you be interested in taking this on?

The tricky parts off the top of my head:

1. Going from `steps_per_execution == 1` to `steps_per_execution > 1` will require clearing the `tf.function` training step and recalling `Model.make_train_function` (All other changes to `steps_per_execution` don't need a retracing)
2. Ensuring the Callbacks and DataHandler are both keeping track of the correct step number

**Comment by gbaned** [comment]
@lgeiger Can you please check @omalleyt12's comments and keep us posted ? Thanks!

**Comment by lgeiger** [comment]
Hi @omalleyt12 thanks for the response, and sorry for my late reply. Do I understand it correctly that the main reason to auto-tune `steps_per_execution` is to have better support for the progressbar?

To me this seems to require quite a bit of code to properly handle all edge cases.  Since `steps_per_execution` gets passed in `compile()` instead of `fit()` autotuning it would effectively make `Model.train_function` not deterministic (if autotuned based on the time per step) which might be problematic if users rely on this to function to run for a fixed number of steps?. I don't think I understand enough of the reasoning behind the design of `steps_per_execution` to properly implement auto-tuning in a short amount of time (e.g. for me intuitively `steps_per_execution` would be an implementation detail of the training process so I'm a bit confused why it is already set in `compile()` and not in fit `fit()`, although that fact probably has a very good reason too :)).

Unfortunately I don't have the time to take another look to properly implement autotuning in the near future so I'd prefer to close this PR if the current solution isn't an option for you.

Personally, I've never really used a batch level progress bar for any performance critical long running trainings, as many remote execution services (including AI Platform last time I checked) don't have great support for handling ANSI carriage returns anyway and when using a Jupyter notebook I usually only train for a very short time during debugging where performance is not really critical (or the model even executes eagerly). So I am not sure yet if I agree with the motivation of autotuning here (although I don't have the full picture over all Keras users, so feel free to ignore this comment).

**Comment by gbaned** [comment]
@lgeiger Can you please resolve conflicts? Thanks!

**Comment by tensorflowbutler** [comment]
It has been 14 days with no activity and the `awaiting response` label was assigned. Is this PR still valid? Assigning the `stalled` label. Please comment to reassure me that this is still being worked on.

**Comment by gbaned** [comment]
@lgeiger  Any update on this PR? Please. Thanks!

**Comment by lgeiger** [comment]
@gbaned I haven't heard back from @omalleyt12 about https://github.com/tensorflow/tensorflow/pull/41742#issuecomment-705246013 so rebasing doesn't make sense before the general question has been resolved.

**Comment by gbaned** [comment]
@omalleyt12  Can you please take a look on the above comment from @lgeiger. Thanks!

**Comment by fchollet** [comment]
Thanks for the PR, and sorry the change didn't work out. Feel free to reopen a new PR if you would still like to see this change (you can also contact me at fchollet@google.com to discuss API & design specifics).

**Comment by lgeiger** [comment]
> you can also contact me at fchollet@google.com to discuss API & design specifics).

Thanks for the offer! Unfortunately, I can't dedicate the time needed to really have a think about possible improvements to the API in the near future. But I will definitely reach to you if I get a chance to look into this again or will provide feedback if there is a public RFC to comment on.
